File: examples/lasso_interop.py
#!/usr/bin/env python3
"""Cross-tool interop harness (lasso side). Companion to examples/lasso_interop.rs.

  python examples/lasso_interop.py read  <path>   # lasso reads scheme A, asserts
  python examples/lasso_interop.py write <path>    # lasso writes scheme B
"""
import sys
import numpy as np
from lasso.dyna import D3plot, ArrayType as AT

NUMNP = 8
NSTATES = 2


def approx(a, b, what):
    a = np.asarray(a, dtype=np.float64)
    b = np.asarray(b, dtype=np.float64)
    assert a.shape == b.shape, f"{what}: shape {a.shape} != {b.shape}"
    assert np.allclose(a, b, atol=1e-4), f"{what}: {a} != {b}"


def read_scheme_a(path):
    d = D3plot(path)
    a = d.arrays
    coords = np.arange(NUMNP * 3, dtype=np.float64).reshape(NUMNP, 3)
    approx(a[AT.node_coordinates], coords, "node_coordinates")

    disp = np.stack([coords + s for s in range(NSTATES)])
    approx(a[AT.node_displacement], disp, "node_displacement")
    approx(a[AT.node_velocity], np.full((NSTATES, NUMNP, 3), 0.5), "node_velocity")

    approx(a[AT.global_kinetic_energy], [10.0, 11.0], "global_kinetic_energy")
    approx(a[AT.global_internal_energy], [20.0, 21.0], "global_internal_energy")
    approx(a[AT.global_total_energy], [30.0, 31.0], "global_total_energy")
    approx(a[AT.global_timesteps], [0.0, 1.0], "global_timesteps")

    # solid: 6 stress + 1 pstrain per state
    ss = np.array([[[1000 * s + v for v in range(6)]] for s in range(NSTATES)], dtype=np.float64)
    approx(a[AT.element_solid_stress].reshape(NSTATES, 1, 6), ss, "element_solid_stress")
    ps = np.array([[1000 * s + 6] for s in range(NSTATES)], dtype=np.float64)
    approx(a[AT.element_solid_effective_plastic_strain].reshape(NSTATES, 1), ps, "solid_pstrain")

    # shell: one layer of 6 stress + 1 pstrain
    sh = np.array([[[2000 * s + v for v in range(6)]] for s in range(NSTATES)], dtype=np.float64)
    approx(a[AT.element_shell_stress].reshape(NSTATES, 1, 6), sh, "element_shell_stress")
    shp = np.array([[2000 * s + 6] for s in range(NSTATES)], dtype=np.float64)
    approx(a[AT.element_shell_effective_plastic_strain].reshape(NSTATES, 1), shp, "shell_pstrain")

    # beams: nv1d = 6 (axial, shear s/t, moment s/t, torsion)
    approx(a[AT.element_beam_axial_force].reshape(NSTATES, 1),
           [[3000 * s + 0] for s in range(NSTATES)], "beam_axial_force")
    approx(a[AT.element_beam_shear_force].reshape(NSTATES, 1, 2),
           [[[3000 * s + 1, 3000 * s + 2]] for s in range(NSTATES)], "beam_shear_force")
    approx(a[AT.element_beam_bending_moment].reshape(NSTATES, 1, 2),
           [[[3000 * s + 3, 3000 * s + 4]] for s in range(NSTATES)], "beam_bending_moment")
    approx(a[AT.element_beam_torsion_moment].reshape(NSTATES, 1),
           [[3000 * s + 5] for s in range(NSTATES)], "beam_torsion_moment")

    # nodal temperature (IT=1)
    temp = np.array([[7000 + 100 * s + n for n in range(NUMNP)] for s in range(NSTATES)], dtype=np.float64)
    approx(a[AT.node_temperature].reshape(NSTATES, NUMNP), temp, "node_temperature")

    print("lasso read scheme A OK: coords/disp/vel, solid+shell stress+pstrain,")
    print("  beam resultants (axial/shear/bending/torsion), node temperature, globals, timesteps")


def write_scheme_b(path):
    d = D3plot()
    a = d.arrays
    coords = np.arange(NUMNP * 3, dtype=np.float32).reshape(NUMNP, 3)
    a[AT.node_coordinates] = coords
    a[AT.node_displacement] = np.stack([coords + s for s in range(NSTATES)]).astype(np.float32)
    a[AT.node_velocity] = np.full((NSTATES, NUMNP, 3), 0.25, dtype=np.float32)
    a[AT.global_timesteps] = np.array([0.0, 1.0], dtype=np.float32)

    # one hex solid over all 8 nodes (lasso node indexes are 0-based)
    a[AT.element_solid_node_indexes] = np.arange(8, dtype=np.int64).reshape(1, 8)
    a[AT.element_solid_part_indexes] = np.array([0], dtype=np.int64)
    a[AT.element_solid_ids] = np.array([1], dtype=np.int64)
    # lasso wants a n_solid_layers axis (index 2): stress (states,solids,layers,6),
    # pstrain (states,solids,layers). One layer -> nv3d = 6 stress + 1 pstrain.
    a[AT.element_solid_stress] = np.array(
        [[[[100 * s + v for v in range(6)]]] for s in range(NSTATES)], dtype=np.float32
    )
    a[AT.element_solid_effective_plastic_strain] = np.array(
        [[[100 * s + 6]] for s in range(NSTATES)], dtype=np.float32
    )

    a[AT.global_kinetic_energy] = np.array([1.0, 2.0], dtype=np.float32)
    a[AT.global_internal_energy] = np.array([3.0, 4.0], dtype=np.float32)
    a[AT.global_total_energy] = np.array([7.0, 8.0], dtype=np.float32)

    d.write_d3plot(path)
    print(f"lasso wrote scheme B -> {path}")


def main():
    if len(sys.argv) != 3 or sys.argv[1] not in ("read", "write"):
        print("usage: lasso_interop.py <read|write> <path>", file=sys.stderr)
        sys.exit(2)
    mode, path = sys.argv[1], sys.argv[2]
    (read_scheme_a if mode == "read" else write_scheme_b)(path)


if __name__ == "__main__":
    main()

File: examples/lasso_interop.rs
//! Cross-tool interop harness for d3plot read/write against open-lasso-python.
//!
//! Usage:
//!   cargo run --example lasso_interop -- write <path>   # our writer  -> file (scheme A)
//!   cargo run --example lasso_interop -- read  <path>   # our reader  <- file (scheme B, asserts)
//!
//! The companion `examples/lasso_interop.py` reads scheme A and writes scheme B,
//! so the two tools validate each other in both directions with matching values.

use dynars::results::{D3plot, D3plotWriter, GlobalField, NodeField, ResultBlock, StateBlock};

// ---- deterministic value schemes shared with the Python side ----
const NUMNP: usize = 8;
const NSTATES: usize = 2;

fn coord(i: usize) -> f64 {
    i as f64 // node n coords = (3n, 3n+1, 3n+2)
}

fn write_scheme_a(path: &str, double: bool) {
    let coords: Vec<f64> = (0..NUMNP * 3).map(coord).collect();
    let mut w = D3plotWriter::new(coords.clone()).unwrap();
    w.set_double_precision(double);
    w.set_title("dynars<->lasso interop A");
    // one hex solid over all 8 nodes, one quad shell over the first 4, one beam.
    w.add_solid([1, 2, 3, 4, 5, 6, 7, 8], 1);
    w.add_shell([1, 2, 3, 4], 2);
    w.add_beam([1, 2, 3], 3);

    // per-element results.
    let mut solid = Vec::new();
    let mut shell = Vec::new();
    let mut beam = Vec::new();
    for s in 0..NSTATES {
        // solids/shells: 7 vars = 6 stress + 1 effective plastic strain.
        for v in 0..7 {
            solid.push((1000 * s + v) as f64);
        }
        for v in 0..7 {
            shell.push((2000 * s + v) as f64);
        }
        // beams: nv1d = 6 (axial, shear s/t, moment s/t, torsion).
        for v in 0..6 {
            beam.push((3000 * s + v) as f64);
        }
    }
    w.set_solid_results(ResultBlock::new([NSTATES, 1, 7], solid));
    w.set_shell_results(ResultBlock::new([NSTATES, 1, 7], shell)); // one layer: 6 stress + pstrain
    w.set_beam_results(ResultBlock::new([NSTATES, 1, 6], beam));

    // global energies.
    w.set_global_history(GlobalField::KineticEnergy, vec![10.0, 11.0]);
    w.set_global_history(GlobalField::InternalEnergy, vec![20.0, 21.0]);
    w.set_global_history(GlobalField::TotalEnergy, vec![30.0, 31.0]);

    // nodal temperature (IT=1).
    let temp: Vec<f64> = (0..NSTATES)
        .flat_map(|s| (0..NUMNP).map(move |n| (7000 + 100 * s + n) as f64))
        .collect();
    w.set_node_field(NodeField::Temperature, temp);

    for s in 0..NSTATES {
        let disp: Vec<f64> = coords.iter().map(|&c| c + s as f64).collect();
        let vel: Vec<f64> = vec![0.5; NUMNP * 3];
        w.add_state(s as f64, disp, Some(vel), None).unwrap();
    }
    w.write(path).unwrap();
    println!("wrote scheme A -> {path}");
}

/// Read a lasso-written file (scheme B) with our reader and assert its contents.
/// Scheme B: NUMNP nodes with coords[i]=i, 2 states, disp = coords + state,
/// velocity = 0.25, one solid [1..8] part 1 with stress 100*s+v (v 0..5) and
/// pstrain 100*s+6, global TE = [7,8].
fn read_scheme_b(path: &str) {
    let d = D3plot::open(path).unwrap();
    let approx = |a: f64, b: f64, what: &str| {
        assert!((a - b).abs() < 1e-4, "{what}: {a} != {b}");
    };
    assert_eq!(d.num_nodes(), NUMNP, "numnp");
    assert_eq!(d.num_states(), NSTATES, "nstates");

    // initial coordinates
    let c0 = d.node_coordinates(0).unwrap();
    for (i, &v) in c0.iter().enumerate() {
        approx(v, i as f64, "coord");
    }
    // displacement (deformed coords) at state 1 = coords + 1
    let c1 = d.node_coordinates(1).unwrap();
    for (i, &v) in c1.iter().enumerate() {
        approx(v, i as f64 + 1.0, "disp state1");
    }
    // velocity block = 0.25 everywhere
    let all = d.resolve_states(None).unwrap();
    let (vel, vdims) = d.block_data(StateBlock::Velocity, &all).unwrap();
    assert_eq!(vdims, [NSTATES, NUMNP, 3], "vel dims");
    for x in vel.to_f64() {
        approx(x, 0.25, "vel");
    }
    // solid raw block = [6 stress, pstrain] per state, values 100*s + v
    let (solid, sdims) = d.block_data(StateBlock::Solid, &all).unwrap();
    assert_eq!(sdims, [NSTATES, 1, 7], "solid dims");
    let sd = solid.to_f64();
    for s in 0..NSTATES {
        for v in 0..7 {
            approx(sd[s * 7 + v], (100 * s + v) as f64, "solid var");
        }
    }
    // global total energy = [7, 8]
    let te = d.global_history(GlobalField::TotalEnergy).unwrap();
    approx(te[0], 7.0, "TE0");
    approx(te[1], 8.0, "TE1");

    println!("read scheme B OK: node coords/disp/vel, solid stress+pstrain, global energy all match");
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let (mode, path) = (args.get(1).map(String::as_str), args.get(2).map(String::as_str));
    match (mode, path) {
        (Some("write"), Some(p)) => write_scheme_a(p, false),
        (Some("write8"), Some(p)) => write_scheme_a(p, true), // double precision
        (Some("read"), Some(p)) => read_scheme_b(p),
        _ => {
            eprintln!("usage: lasso_interop <write|write8|read> <path>");
            std::process::exit(2);
        }
    }
}
